[stdlib] Adds functionality to spawn and manage processes from exec. file

izo0x90 opened the pull request:
[Foundation for this PR is set here]( https://github.com/modular/mojo/pull/4017), it adds the needed lowlevel utilities

This PR:
- Adds utility struct. to allow spawning and managing child processes
  from an executable file
  - `run` factory method that creates process and returns "management"
    object
  - Instance methods to send signals to child process, INT, HUP, KILL
- Adds utility struct. to handle Posix pipe interprocess communication

Ex.:
```
import os

fn main():
 p = os.Process.run("ls", ["-lha"])
 p.interrupt()
```

Example usage:
https://github.com/izo0x90/mojo/blob/b99503e7bc5b2f210c957e5cbd8d0fb8b2c9ed1a/examples/process/process_example.mojo#L22-L146

izo0x90 committed:
Adds functionality to spawn and manage processes from exec. file

- Adds fork, execvp, kill system call utils. to Mojos cLib binds

- Adds utility struct. to allow spawning and managing child processes
  from an executable file
  - `run` factory method that creates process and returns "management"
    object
  - Instance methods to send signals to child process, INT, HUP, KIL

Signed-off-by: Hristo I. Gueorguiev <53634432+izo0x90@users.noreply.github.com>

izo0x90 committed:
Fixes/ updates to `process` functionality

- Move `process` related code to its own file
- Use `vfork` to avoid issues with threading until they are resolved in
  Mojo
- Sending signals to process returns success status
- Docstr additions

Signed-off-by: Hristo I. Gueorguiev <53634432+izo0x90@users.noreply.github.com>

owenhilyard reviewed (commented):
Good work so far, just a few fixes.

izo0x90 committed:
Use interprocess pipe to signal exec error in forked process

Implements canonical fork/ exec error signaling process that uses a pipe
configure to close on `exec` sys. call to notify the parent process if
an error occurs in the child and it is unable to `exec` the desired
executable file

- Implement Pipe abstraction
- Add `read_bytes` capability to `FileDecscriptor`
- Add file descriptor controls function to Libc. bindings
- Fix previously broken test for rasing on exec errors

Refs:
- https://cr.yp.to/docs/selfpipe.html

Signed-off-by: Hristo I. Gueorguiev <53634432+izo0x90@users.noreply.github.com>

izo0x90 committed:
Update `process.Pipe` to take in data buffer

- `read_bytes` takes in already alloced. buffer to avoid n+1 issues in
  loops etc.

Signed-off-by: Hristo I. Gueorguiev <53634432+izo0x90@users.noreply.github.com>

izo0x90 committed:
Update file_descriptor `read_bytes` internals

Update the list creation to conform to upcoming List API changes that
make the list size a `private attribute`

Signed-off-by: Hristo I. Gueorguiev <53634432+izo0x90@users.noreply.github.com>

izo0x90 committed:
Update `file_descriptor.read_bytes` do not alloc.

- `read_bytes` takes in Span[Byte] buffer as opposed to allocating
  to avoid n+1 alloc. issues etc.

- Remove c_str_ptr type from ffi module since it is not a convention set
  by C lang., it will potentially get added back in at a later date
  after discussions and under its own PR

- Doc string updates to libC funcs.

Co-authored-by: Connor Gray <accounts@connorgray.com>
Signed-off-by: Hristo (Izo) G. <53634432+izo0x90@users.noreply.github.com>

JoeLoser commented on stdlib/test/os/test_process.mojo:
```diff
@@ -0,0 +1,45 @@
+
+def main():
+    test_process_run()
+    # TODO: How can exception being raised on missing file be asserted
```
I hope to have some time next week to take a look more into this.  The team has been really busy working on some things we're showcasing next week at Nvidia GTC - sorry for the delay!

izo0x90 replied:
No worries at all, thank you, let me know what I can do once you guys get a chance to take a look 🔥 ... Best of luck with the showcase !!!!

izo0x90 committed:
Merge branch 'main' of github-personal:izo0x90/mojo into Hristo/Add-basic-process-exec-facilities

izo0x90 committed:
Update code to work with current mojo syntax and conventions

izo0x90 committed:
Merge branch 'main' of github-personal:izo0x90/mojo into Hristo/Add-basic-process-exec-facilities

izo0x90 committed:
Merge branch 'main' into Hristo/Add-basic-process-exec-facilities

izo0x90 committed:
Fixes for exec error checking

- Remove heap allocation for error message read from pipe, we are the
  one setting it so length is compile time known, no need for allocations.

- Raise and error if we fail to read from the pipe

izo0x90 committed:
Merge branch 'Hristo/Add-basic-process-exec-facilities' of github-personal:izo0x90/mojo into Hristo/Add-basic-process-exec-facilities

izo0x90 committed:
Merge branch 'main' of github-personal:izo0x90/mojo into Hristo/Add-basic-process-exec-facilities

izo0x90 committed:
Formatting fixes and remove unneeded change in _libc

- Unneeded change to origin parameter in _libc, removing
- Fix code formatting

izo0x90 commented on mojo/stdlib/stdlib/os/process.mojo:
```diff
@@ -0,0 +1,284 @@
+                argv_array_ptr_cstr_ptr[offset] = file_name.unsafe_cstr_ptr()
+                offset += 1
+
+                for var arg in argv:
```
We need to copy the strings here because `unsafe_cstr_ptr` is "mutating". Could use `ref` but then we would require the args list to be passed in as `mut`.

izo0x90 commented on mojo/stdlib/stdlib/os/process.mojo:
```diff
@@ -0,0 +1,284 @@
+                err = StringSlice(unsafe_from_utf8=buf)
+            except e:
+                err = None
+
```
+1 to the LLMs lol. No need for an allocation here at all actually, since we are the ones setting the error string which never changes all lengths are know at compile time so we can just use an InlineArray and avoid the heap altogether.

izo0x90 committed:
Merge branch 'main' into Hristo/Add-basic-process-exec-facilities

barcharcraz commented on mojo/stdlib/std/os/process.mojo:
```diff
@@ -0,0 +1,290 @@
+
+    fn __init__(
+        out self,
+        in_close_on_exec: Bool = False,
+        out_close_on_exec: Bool = False,
```
IMO the default should be true.

barcharcraz commented on mojo/stdlib/stdlib/os/process.mojo:
```diff
@@ -0,0 +1,290 @@
+
+            if pid == 0:
+                # Child process.
+                pipe.set_output_only()
```
nb: this is only OK because pipe is on the stack. Would prefer we just close the fd and not write onto the stack.

barcharcraz commented on mojo/stdlib/stdlib/os/process.mojo:
```diff
@@ -0,0 +1,290 @@
+                _ = execvp(path_cptr, argv_array_ptr_cstr_ptr)
+
+                # This will only get reached if exec call fails to replace currently executing code
+                argv_array_ptr_cstr_ptr.free()
```
yes, and because of that I'm pretty sure the allocation leaks in the parent process. vfork doesn't change the page-table mapping, so the allocation is in BOTH processes and it's freed ONLY in the child by virtue of the exec wiping everything out.

barcharcraz commented on mojo/stdlib/stdlib/os/process.mojo:
```diff
@@ -0,0 +1,290 @@
+                argv_array_ptr_cstr_ptr.free()
+
+                # Canonical fork/ exec error handling pattern of using a pipe that closes on exec is
+                # used to signal error to parent process `https://cr.yp.to/docs/selfpipe.html`
```
Not sure about macOS but on linux you can use `CLONE_PIDFD` to avoid this race

barcharcraz replied:
and this is another reason to just call `posix_spawn`

barcharcraz commented on mojo/stdlib/stdlib/os/process.mojo:
```diff
@@ -0,0 +1,290 @@
+                # used to signal error to parent process `https://cr.yp.to/docs/selfpipe.html`
+                pipe.write_bytes(exec_err_code.as_bytes())
+
+                exit(1)
```
I think this will call the parent's atexit handlers and flush the parents io buffers

barcharcraz commented on mojo/stdlib/stdlib/os/process.mojo:
```diff
@@ -0,0 +1,290 @@
+                    + String(e)
+                )
+
+            if err and len(err.value()) > 0 and err.value() == exec_err_code:
```
does the child need to be reaped in this case?

izo0x90 committed:
Merge branch 'main' of github-personal:izo0x90/mojo into Hristo/Add-basic-process-exec-facilities

izo0x90 committed:
First steps to using posix_spawnp

izo0x90 committed:
Finalize Process handling using

izo0x90 committed:
V1 of updating os/process tests

izo0x90 committed:
Finalize v1 process module

izo0x90 committed:
Merge branch 'main' into Hristo/Add-basic-process-exec-facilities

izo0x90 committed:
Update process module related code to newst syntax/ semantics

izo0x90 committed:
Fix process module doc strings

izo0x90 committed:
Remove no longer needed constant

izo0x90 committed:
Formatting for doc string got me

izo0x90 committed:
Merge branch 'main' into Hristo/Add-basic-process-exec-facilities

izo0x90 committed:
Lib C formatting

izo0x90 committed:
Fix process examples bazel config imports

barcharcraz commented on mojo/stdlib/std/os/process.mojo:
```diff
@@ -0,0 +1,441 @@
+        Raises:
+            Error: If the pipe could not be created or configured.
+        """
+        var pipe_fds = alloc[c_int](2)
```
must we malloc these? I think we can use an inline array.

barcharcraz commented on mojo/stdlib/std/os/process.mojo:
```diff
@@ -0,0 +1,441 @@
+            raise Error("Failed to create pipe")
+
+        if in_close_on_exec:
+            if not self._set_close_on_exec(pipe_fds[0]):
```
Ideally we'd call `pipe2` to avoid the race where some other thread execs in between the `pipe` and the `fcntl`, but macOS doesn't have pipe2 and avoiding these races is a bit of a sisyphean task.

barcharcraz reviewed (commented):
This looks good aside from the allocation in `Pipe` (which honestly is minor, I can correct it while merging if you'd like).

One annoyance is we have no way of detaching a process, but this is complicated and can be left for later. I think it should just be a matter of defining flags for `posix_spawn` but they'll be somewhat platform specific.

JoeLoser commented on mojo/stdlib/std/os/process.mojo:
```diff
@@ -0,0 +1,441 @@
+        try:
+            _ = self.wait()
+        except:
+            # Errors in __del__ should be suppressed.
```
**Question** If `wait()` fails (e.g. already called, or some other error) - the child process may become a zombie. Is there anything better we can do here @barcharcraz? Suppressing all errors silently can mask real issues.

JoeLoser commented on mojo/stdlib/std/os/process.mojo:
```diff
@@ -0,0 +1,441 @@
+            # Errors in __del__ should be suppressed.
+            pass
+
+    fn _kill(mut self, signal: Int) -> Bool:
```
This is a TOCTOU (time of check, time of use) race I believe (right @barcharcraz?)

Like your comment says, the process can exit between `poll()` and `kill()`, potentially allowing the PID to be reused for a different process. I'd recommend we just call `kill()` directly and check for `ESRCH` errno. The kernel handles this automatically for us:

```mojo
fn _kill(mut self, signal: Int) -> Bool:
  if kill(self.child_pid, signal) == 0:
      return True
  var err = get_errno()
  return err != ESRCH # False if process doesn't exist, True for other errors
```

barcharcraz replied:
The PID won't be reused even if the process has already exited. The process will become a zombie until reaped because we don't change the default signal disposition for sigchld.

The way SIGCHLD works is confusing, if you set the disposition to SIG_IGN then the process isn't zombified at all (and even the poll might poll the wrong process), but if it's SIG_DFL then you _do_ get zombies even though the "default" action is to ignore the signal.

Anyway, since we aren't messing with SIGCHLD we do get zombies. Either we poll and the process has exited, in which case the poll will reap the process and we never kill, or it hasn't and the process isn't reaped. After being killed it will be zombified to be handled by a later `wait`/`pool`/`__del__`

barcharcraz replied:
Turns out it's hard to kill your children :)

JoeLoser commented on mojo/stdlib/std/os/process.mojo:
```diff
@@ -0,0 +1,441 @@
+        self.fd_out = FileDescriptor(Int(pipe_fds[1]))
+        pipe_fds.free()
+
+    fn __del__(deinit self):
```
I think we have some resource management issues here.  `close()` can fail (e.g. `EINTR`, `EIO`) but the errors are ignored here.  This may be easier to ensure correctness if we initialize `fd_in` and `fd_out` to `None` before any operations that raise in `__init__` for example. Then we could check the return values from `close()` and handle errors appropriately.

barcharcraz replied:
It's actually OK to swallow errors from close. EBADF is basically structurally impossible, historically everybody except HP-UX would close the fd before returning EINTR, creating inconsistency. Now the open group has said the FD _MUST_ still be open if EINTR is returned but EINPROGRESS should indicate the call was interrupted but the fd was closed.

Sadly this change needs to be adopted by linux. Basically everyone still does the linux semantics.

I don't think we need to care about HP-UX, we can assume the sane linux semantics.

For other IO errors there's nothing we can do by the time we get to this close call. If the user wanted to handle them they would have synched the file.


See:
https://lwn.net/Articles/576478/
https://ewontfix.com/4/

owenhilyard replied:
@barcharcraz Can you include that last comment inline in the function (no need for a docstring) as a note for future readers? If Linux ever updates its behavior (unlikely) we'll need to fix that, and there's a non-zero chance of another POSIX OS (Redox comes to mind) being written to the spec with the now-correct behavior. Having that information there will save a lot of time.

JoeLoser commented on mojo/stdlib/std/os/process.mojo:
```diff
@@ -0,0 +1,441 @@
+
+            var pid: c_pid_t = 0
+
+            var has_error_code = posix_spawnp(
```
**Note** `posix_spawnp` call here passes the environment pointer as NULL, meaning the child inherits the parent's environment. Currently we provide no way to customize this: is that intentional?

**Suggestion**: We could add an optional `env: Optional[Dict[String, String]]` argument to `run()` and use it here.

izo0x90 replied:
Yes, my thinking here was to add this as a follow up... was't sure what the best way to add a todo was.

JoeLoser commented on mojo/stdlib/std/os/process.mojo:
```diff
@@ -0,0 +1,441 @@
+            self.status = result
+        return result
+
+    fn wait(mut self) raises -> ProcessStatus:
```
**Question** What happens if you call `wait()` multiple times? The code currently caches the status (which is good) but this isn't documented. I'd suggest we document that `wait()` and `poll()` cache results so that subsequent calls return the cached status.

JoeLoser commented on mojo/stdlib/std/os/process.mojo:
```diff
@@ -0,0 +1,441 @@
+    fn set_input_only(mut self):
+        """Close the output descriptor/ channel for this side of the pipe."""
+        if self.fd_out:
+            _ = close(rebind[Int](self.fd_out.value()))
```
**Question** Why is the `rebind[Int]` needed here and below in `set_output_only`, etc? By using `rebind`, we're bypassing type checking essentially. Could we make it so that `FileDescriptor` can be passed directly to `close()` perhaps?

barcharcraz replied:
I think we should have a higher level FD api. For now I'd kinda rather just use c_int directly throughout (Int is actually wrong, anyway).

barcharcraz replied:
Honestly, Pipe kinda _is_ the higher level fd api.

JoeLoser commented on mojo/stdlib/std/os/process.mojo:
```diff
@@ -0,0 +1,441 @@
+            constrained[
+                False, "Unknown platform process execution not implemented"
+            ]()
+            abort[prefix="ERROR:"](
```
**Question** Is this reachable, or is it redundant?

izo0x90 replied:
A return is expected out of the function ... so even tho. we have constrained in this code path the compiler still complains about not having a return if we do not explicitly call abort

JoeLoser commented on mojo/stdlib/std/os/process.mojo:
```diff
@@ -0,0 +1,441 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ===----------------------------------------------------------------------=== #
+"""Implements os methods for dealing with processes.
```
It'd be great to add some module/API docstrings for some of the more nuanced things for usage patterns, caveats, etc - especially around resource cleanup.

1. What happens if you call signal methods after the process has exited?
2. Are these methods thread-safe?
3. What's the behavior of calling `wait()` multiple times?
4. Should users always call `wait()` or is relying on `__del__` sufficient?

JoeLoser commented on mojo/stdlib/std/os/process.mojo:
```diff
@@ -0,0 +1,441 @@
+                    + String(has_error_code)
+                )
+
+            argv_array_ptr_cstr_ptr.free()
```
**Suggestion** Good that we free the `argv` stuff, but can we document the lifetime requirements of the `String` objects themselves in `argv`?

izo0x90 replied:
What is the best way to document that ?

JoeLoser commented on mojo/stdlib/test/os/test_process_stdout.mojo:
```diff
@@ -0,0 +1,53 @@
+# ===----------------------------------------------------------------------=== #
```
**Suggestion** I'd like to see some comprehensive tests, especially for `Pipe` which has zero test coverage right now.

Some ideas of tests that are missing right now:

- Exit code verification
- Signal handling (interrupt, hangup, kill)
- Polling vs. waiting
- Multiple wait calls (we have this in an example elsewhere though IIRC)
- Pipe functionality is completely untested right now
- Process that exits immediately
- Process that runs for a long time (this is OK though as you have an example elsewhere)

JoeLoser commented on mojo/stdlib/std/os/process.mojo:
```diff
@@ -0,0 +1,441 @@
+        return result
+
+    @staticmethod
+    fn run(var path: String, argv: List[String]) raises -> Process:
```
**Suggestion** Currently, we don't have a way to capture stdout/stderr from the child, provide stdin to the child, or redirect I/O to files or pipes.  This would all be good essential functionality for a process API. I'd suggest tackling this in a follow-up by adding optional `stdin`, `stdout`, and `stderr` arguments that accept `Pipe` or file descriptors.

owenhilyard replied:
Being able to supply `String`/`StringSlice`/`StringLiteral` would be nice as well.

izo0x90 replied:
Is there a way to do that in a "Trait" way somehow yet ? Or do we just do a union type of those 3 ?

owenhilyard replied:
I think you'll need to do it with function overloading and wrap the string versions around the FD versions.

JoeLoser reviewed (commented):
This is looking good! I left a few comments to check out before we land this.  Feel free to consider some of them as things to tackle as follow-ups - use your best judgement and work with @barcharcraz on it.

Don't forget to also add a [changelog entry here](https://github.com/modular/modular/blob/13736a8166b257e09dadcdb08ff73baf36084de5/mojo/docs/changelog.md#library-changes) so others know about your great improvements!

Thanks again for working on this, very exciting to see!

izo0x90 committed:
Add stdio redirects and env to process module

izo0x90 committed:
Merge branch 'main' into Hristo/Add-basic-process-exec-facilities

izo0x90 committed:
Naming consistency in process module examples

izo0x90 committed:
Better resource management in Process module

- Pipe does not need to dynamically alloc, use InlineArray
- RAII approach to Process args, env and file actions resource
  allocation to make sure deallocations run across all failure paths
